File: app/src/main/java/com/example/mixmate/SettingsActivity.java
package com.example.mixmate;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.google.android.material.card.MaterialCardView;
import com.google.android.material.materialswitch.MaterialSwitch;

public class SettingsActivity extends AppCompatActivity {

    private static final String TAG = "SettingsActivity";
    private static final String PREFS_NAME = "app_preferences";
    private static final String KEY_PUSH_NOTIFICATIONS = "push_notifications";
    private static final String KEY_RECIPE_UPDATES = "recipe_updates";
    private static final String KEY_MEASUREMENT_UNITS = "measurement_units";
    private static final String KEY_LANGUAGE = "selected_language";

    // UI Components
    private ImageButton btnBack;
    private MaterialCardView cardEditProfile, cardChangePassword, cardUnits, cardLanguage, cardPrivacy,
            cardHelpSupport, cardAbout, cardLogout;
    private MaterialSwitch switchPushNotifications, switchRecipeUpdates;
    private TextView tvCurrentUnits, tvCurrentLanguage;

    // Preferences
    private SharedPreferences sharedPreferences;

    @Override
    protected void attachBaseContext(android.content.Context newBase) {
        super.attachBaseContext(LocaleHelper.INSTANCE.onAttach(newBase));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        EdgeToEdge.enable(this);
        setContentView(R.layout.activity_settings);
        
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
            var systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
            return insets;
        });

        // Initialize preferences
        sharedPreferences = getSharedPreferences(PREFS_NAME, MODE_PRIVATE);

        initializeViews();
        setupClickListeners();
        loadPreferences();

        Log.d(TAG, "Settings activity initialized");
    }


    private void initializeViews() {
        try {
            // Navigation
            btnBack = findViewById(R.id.btn_back);

            // Account section
            cardEditProfile = findViewById(R.id.card_edit_profile);
            cardChangePassword = findViewById(R.id.card_change_password);

            // Preferences section
            cardUnits = findViewById(R.id.card_units);
            tvCurrentUnits = findViewById(R.id.tv_current_units);
            cardLanguage = findViewById(R.id.card_language);
            tvCurrentLanguage = findViewById(R.id.tv_current_language);

            // Notifications section
            switchPushNotifications = findViewById(R.id.switch_push_notifications);
            switchRecipeUpdates = findViewById(R.id.switch_recipe_updates);

            // Privacy & Support section
            cardPrivacy = findViewById(R.id.card_privacy);
            cardHelpSupport = findViewById(R.id.card_help_support);
            cardAbout = findViewById(R.id.card_about);

            // Account actions
            cardLogout = findViewById(R.id.card_logout);

            Log.d(TAG, "All views initialized successfully");
        } catch (Exception e) {
            Log.e(TAG, "Error initializing views", e);
            Toast.makeText(this, getString(R.string.toast_error_init_settings), Toast.LENGTH_SHORT).show();
        }
    }

    private void setupClickListeners() {
        // Back button
        btnBack.setOnClickListener(v -> {
            Log.d(TAG, "Back button clicked");
            finish();
        });

        // Account section
        cardEditProfile.setOnClickListener(v -> {
            Log.d(TAG, "Edit Profile clicked");
            try {
                Intent intent = new Intent(this, EditProfileActivity.class);
                startActivity(intent);
            } catch (Exception e) {
                Log.e(TAG, "Error starting EditProfileActivity: " + e.getMessage());
                Toast.makeText(this, getString(R.string.toast_error_open_edit_profile), Toast.LENGTH_SHORT).show();
            }
        });

        cardChangePassword.setOnClickListener(v -> {
            Log.d(TAG, "Change Password clicked");
            Toast.makeText(this, getString(R.string.toast_change_password_coming), Toast.LENGTH_SHORT).show();
            // TODO: Navigate to ChangePasswordActivity
        });

        // Preferences section
        cardUnits.setOnClickListener(v -> {
            Log.d(TAG, "Measurement Units clicked");
            showUnitsDialog();
        });

        cardLanguage.setOnClickListener(v -> {
            Log.d(TAG, "Language clicked");
            showLanguageDialog();
        });


        // Notifications section
        switchPushNotifications.setOnCheckedChangeListener((buttonView, isChecked) -> {
            Log.d(TAG, "Push notifications switched: " + isChecked);
            saveBooleanPreference(KEY_PUSH_NOTIFICATIONS, isChecked);
            String message = isChecked ? getString(R.string.toast_push_enabled) : getString(R.string.toast_push_disabled);
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
        });

        switchRecipeUpdates.setOnCheckedChangeListener((buttonView, isChecked) -> {
            Log.d(TAG, "Recipe updates switched: " + isChecked);
            saveBooleanPreference(KEY_RECIPE_UPDATES, isChecked);
            String message = isChecked ? getString(R.string.toast_recipe_updates_enabled) : getString(R.string.toast_recipe_updates_disabled);
            Toast.makeText(this, message, Toast.LENGTH_SHORT).show();
        });

        // Privacy & Support section
        cardPrivacy.setOnClickListener(v -> {
            Log.d(TAG, "Privacy Policy clicked");
            openPrivacyPolicy();
        });

        cardHelpSupport.setOnClickListener(v -> {
            Log.d(TAG, "Help & Support clicked");
            openHelpSupport();
        });

        cardAbout.setOnClickListener(v -> {
            Log.d(TAG, "About clicked");
            showAboutDialog();
        });

        // Account actions
        cardLogout.setOnClickListener(v -> {
            Log.d(TAG, "Logout clicked");
            showLogoutDialog();

        });
    }

    private void loadPreferences() {
        try {

            // Load Notifications preferences
            boolean pushNotifications = sharedPreferences.getBoolean(KEY_PUSH_NOTIFICATIONS, true);
            switchPushNotifications.setChecked(pushNotifications);

            boolean recipeUpdates = sharedPreferences.getBoolean(KEY_RECIPE_UPDATES, false);
            switchRecipeUpdates.setChecked(recipeUpdates);

            // Load Measurement Units preference
            String units = sharedPreferences.getString(KEY_MEASUREMENT_UNITS, getString(R.string.units_metric));
            tvCurrentUnits.setText(units);

            // Load Language preference
            String languageCode = LocaleHelper.INSTANCE.getLanguage(this);
            String languageName = languageCode.equals("af") ? getString(R.string.language_afrikaans) : getString(R.string.language_english);
            tvCurrentLanguage.setText(languageName);

            Log.d(TAG, "Preferences loaded successfully");
        } catch (Exception e) {
            Log.e(TAG, "Error loading preferences", e);
            Toast.makeText(this, getString(R.string.toast_error_loading_settings), Toast.LENGTH_SHORT).show();
        }
    }

    private void saveBooleanPreference(String key, boolean value) {
        try {
            sharedPreferences.edit().putBoolean(key, value).apply();
            Log.d(TAG, "Preference saved: " + key + " = " + value);
        } catch (Exception e) {
            Log.e(TAG, "Error saving preference: " + key, e);
            Toast.makeText(this, getString(R.string.toast_error_saving_setting), Toast.LENGTH_SHORT).show();
        }
    }

    private void saveStringPreference(String key, String value) {
        try {
            sharedPreferences.edit().putString(key, value).apply();
            Log.d(TAG, "Preference saved: " + key + " = " + value);
        } catch (Exception e) {
            Log.e(TAG, "Error saving preference: " + key, e);
            Toast.makeText(this, getString(R.string.toast_error_saving_setting), Toast.LENGTH_SHORT).show();
        }
    }

    private void showUnitsDialog() {
        String[] units = {getString(R.string.units_metric), getString(R.string.units_imperial), getString(R.string.units_both)};
        String currentUnits = sharedPreferences.getString(KEY_MEASUREMENT_UNITS, getString(R.string.units_metric));

        int selectedIndex = 0;
        for (int i = 0; i < units.length; i++) {
            if (units[i].equals(currentUnits)) {
                selectedIndex = i;
                break;
            }
        }

        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.units_dialog_title))
                .setSingleChoiceItems(units, selectedIndex, (dialog, which) -> {
                    String selectedUnits = units[which];
                    saveStringPreference(KEY_MEASUREMENT_UNITS, selectedUnits);
                    tvCurrentUnits.setText(selectedUnits);
                    Toast.makeText(this, getString(R.string.units_changed_to, selectedUnits), Toast.LENGTH_SHORT).show();
                    dialog.dismiss();
                })
                .setNegativeButton(getString(R.string.dialog_cancel), (dialog, which) -> dialog.dismiss())
                .show();
    }

    private void showLanguageDialog() {
        String[] languages = {getString(R.string.language_english), getString(R.string.language_afrikaans)};
        String[] languageCodes = {"en", "af"};

        String currentLanguage = LocaleHelper.INSTANCE.getLanguage(this);

        int selectedIndex = 0;
        for (int i = 0; i < languageCodes.length; i++) {
            if (languageCodes[i].equals(currentLanguage)) {
                selectedIndex = i;
                break;
            }
        }

        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.select_language))
                .setSingleChoiceItems(languages, selectedIndex, (dialog, which) -> {
                    String selectedLanguageCode = languageCodes[which];
                    String selectedLanguageName = languages[which];

                    // Save language preference
                    LocaleHelper.INSTANCE.setLanguage(this, selectedLanguageCode);

                    // Update UI
                    tvCurrentLanguage.setText(selectedLanguageName);

                    Log.d(TAG, "Language changed to: " + selectedLanguageName);

                    // Recreate activity to apply language change
                    recreate();

                    dialog.dismiss();
                })
                .setNegativeButton(getString(R.string.dialog_cancel), (dialog, which) -> dialog.dismiss())
                .show();
    }


    private void openPrivacyPolicy() {
        try {
            // In a real app, this would be your actual privacy policy URL
            String url = "https://example.com/privacy-policy";
            Intent intent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
            startActivity(intent);
        } catch (Exception e) {
            Log.e(TAG, "Error opening privacy policy", e);
            Toast.makeText(this, getString(R.string.toast_privacy_policy_coming), Toast.LENGTH_SHORT).show();
        }
    }

    private void openHelpSupport() {
        try {
            // Create email intent for support
            Intent emailIntent = new Intent(Intent.ACTION_SENDTO);
            emailIntent.setData(Uri.parse("mailto:support@mixmate.com"));
            emailIntent.putExtra(Intent.EXTRA_SUBJECT, "MixMate Support Request");
            emailIntent.putExtra(Intent.EXTRA_TEXT, "Hi MixMate Team,\n\nI need help with:\n\n");
            
            if (emailIntent.resolveActivity(getPackageManager()) != null) {
                startActivity(emailIntent);
            } else {
                Toast.makeText(this, getString(R.string.toast_no_email_app), Toast.LENGTH_SHORT).show();
            }
        } catch (Exception e) {
            Log.e(TAG, "Error opening help & support", e);
            Toast.makeText(this, getString(R.string.toast_help_support_coming), Toast.LENGTH_SHORT).show();
        }
    }

    private void showAboutDialog() {
        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.about_mixmate))
                .setMessage("MixMate v1.0.0\n\n" +
                           "The ultimate cocktail companion app.\n\n" +
                           "Discover, create, and share amazing cocktail recipes.\n\n" +
                           "© 2024 MixMate Team")
                .setPositiveButton(getString(R.string.dialog_ok), (dialog, which) -> dialog.dismiss())
                .setNeutralButton("Rate App", (dialog, which) -> {
                    // In a real app, this would open the Play Store
                    Toast.makeText(this, getString(R.string.toast_thanks_rating_coming), Toast.LENGTH_SHORT).show();
                })
                .show();
    }

    private void showLogoutDialog() {
        new AlertDialog.Builder(this)
                .setTitle(getString(R.string.dialog_logout_title))
                .setMessage(getString(R.string.dialog_logout_message))
                .setPositiveButton(getString(R.string.logout), (dialog, which) -> {
                    performLogout();
                })
                .setNegativeButton(getString(R.string.dialog_cancel), (dialog, which) -> dialog.dismiss())
                .show();
    }

    private void performLogout() {
        try {
            Log.d(TAG, "Starting logout process");
            
            // Sign out user from Firebase and clear local data using UserManager
            UserManager.INSTANCE.signOut(this);
            
            Log.d(TAG, "UserManager.signOut() completed");
            
            // Additional verification - check if user is still logged in
            boolean stillLoggedIn = UserManager.INSTANCE.isLoggedIn(this);
            Log.d(TAG, "User still logged in after signOut: " + stillLoggedIn);
            
            // Show logout message
            Toast.makeText(this, getString(R.string.toast_logged_out), Toast.LENGTH_SHORT).show();

            Log.d(TAG, "Navigating to MainActivity (login screen)");
            
            // Navigate to MainActivity (login screen) and clear the back stack
            Intent intent = new Intent(this, MainActivity.class);
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
            
            // Add extra flag to ensure complete task clearing
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
            
            // Add extra data to indicate this is a logout scenario
            intent.putExtra("logout", true);
            
            startActivity(intent);
            finish();
            
            Log.d(TAG, "Logout process completed - should now be on login screen");
        } catch (Exception e) {
            Log.e(TAG, "Error during logout process", e);
            Toast.makeText(this, getString(R.string.toast_error_logout, e.getMessage()), Toast.LENGTH_LONG).show();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        // Reload preferences in case they were changed elsewhere
        loadPreferences();
    }
}

File: app/src/main/java/com/example/mixmate/SettingsActivity.kt
package com.example.mixmate

import android.Manifest
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.widget.ImageButton
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.ContextCompat
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import com.example.mixmate.notifications.NotificationHelper
import com.example.mixmate.notifications.NotificationScheduler
import com.google.android.material.card.MaterialCardView
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.materialswitch.MaterialSwitch
import com.google.firebase.auth.FirebaseAuth
import android.widget.TextView
import android.util.Log

/**
 * Settings Activity - Handles all app settings per POE Part 3 requirements:
 * - Theme preferences (dark/light/system)
 * - Language selection (multi-language support)
 * - Notification preferences
 * - Account management (sign out, delete account)
 * - Privacy policy & support links
 */
class SettingsActivity : AppCompatActivity() {

    private lateinit var prefs: SharedPreferences
    private lateinit var btnBack: ImageButton
    
    // Account Section
    private lateinit var cardEditProfile: MaterialCardView
    private lateinit var cardChangePassword: MaterialCardView
    
    // Preferences Section
    private lateinit var cardTheme: MaterialCardView
    private lateinit var tvCurrentTheme: TextView
    private lateinit var cardLanguage: MaterialCardView
    private lateinit var tvCurrentLanguage: TextView
    private lateinit var cardUnits: MaterialCardView
    private lateinit var tvCurrentUnits: TextView
    
    // Notifications Section
    private lateinit var switchPushNotifications: MaterialSwitch
    private lateinit var switchRecipeUpdates: MaterialSwitch
    private lateinit var switchDailyFacts: MaterialSwitch

    // Privacy & Support Section
    private lateinit var cardPrivacy: MaterialCardView
    private lateinit var cardHelpSupport: MaterialCardView
    private lateinit var cardAbout: MaterialCardView
    
    // Account Actions
    private lateinit var cardLogout: MaterialCardView

    companion object {
        private const val PREFS_NAME = "MixMateSettings"
        const val KEY_THEME = "theme_mode"
        const val KEY_LANGUAGE = "app_language"
        const val KEY_UNITS = "measurement_units"
        const val KEY_PUSH_NOTIFICATIONS = "push_notifications"
        const val KEY_RECIPE_UPDATES = "recipe_updates"
        const val KEY_DAILY_FACTS = "daily_cocktail_facts"

        private const val NOTIFICATION_PERMISSION_REQUEST_CODE = 1001

        // Theme modes
        const val THEME_DARK = "dark"
        const val THEME_LIGHT = "light"
        const val THEME_SYSTEM = "system"
        
        // Languages (POE requirement: 2+ South African languages)
        const val LANG_ENGLISH = "en"
        const val LANG_AFRIKAANS = "af"
        const val LANG_ZULU = "zu"
        
        // Units
        const val UNITS_METRIC = "metric"
        const val UNITS_IMPERIAL = "imperial"
    }

    // Permission launcher for notifications (Android 13+)
    private val requestNotificationPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted ->
        if (isGranted) {
            // Permission granted, create channel then enable daily facts
            NotificationHelper.createNotificationChannel(this)
            enableDailyFacts()
        } else {
            // Permission denied, disable switch
            switchDailyFacts.isChecked = false
            Toast.makeText(
                this,
                "Notification permission is required for daily cocktail facts",
                Toast.LENGTH_LONG
            ).show()
        }
    }

    private val requestGenericNotificationPermissionLauncher = registerForActivityResult(
        ActivityResultContracts.RequestPermission()
    ) { isGranted ->
        if (pendingEnablePush && !isGranted) {
            switchPushNotifications.isChecked = false
            Toast.makeText(this, "Notification permission required", Toast.LENGTH_SHORT).show()
        } else if (pendingEnableRecipe && !isGranted) {
            switchRecipeUpdates.isChecked = false
            Toast.makeText(this, "Notification permission required", Toast.LENGTH_SHORT).show()
        } else if (isGranted) {
            // Create channel once permission granted
            NotificationHelper.createNotificationChannel(this)
            if (pendingEnablePush) {
                prefs.edit().putBoolean(KEY_PUSH_NOTIFICATIONS, true).apply()
                Toast.makeText(this, "Push notifications enabled", Toast.LENGTH_SHORT).show()
            }
            if (pendingEnableRecipe) {
                prefs.edit().putBoolean(KEY_RECIPE_UPDATES, true).apply()
                Toast.makeText(this, "Recipe updates enabled", Toast.LENGTH_SHORT).show()
            }
        }
        pendingEnablePush = false
        pendingEnableRecipe = false
    }

    private var pendingEnablePush = false
    private var pendingEnableRecipe = false

    override fun attachBaseContext(newBase: Context) {
        super.attachBaseContext(LocaleHelper.onAttach(newBase))
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        
        // Apply window insets
        ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main)) { v, insets ->
            val systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars())
            v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom)
            insets
        }
        
        prefs = getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
        
        initializeViews()
        loadCurrentSettings()
        setupClickListeners()
    }

    private fun initializeViews() {
        btnBack = findViewById(R.id.btn_back)
        
        // Account
        cardEditProfile = findViewById(R.id.card_edit_profile)
        cardChangePassword = findViewById(R.id.card_change_password)
        
        // Preferences
        cardTheme = findViewById(R.id.card_theme)
        tvCurrentTheme = findViewById(R.id.tv_current_theme)
        cardLanguage = findViewById(R.id.card_language)
        tvCurrentLanguage = findViewById(R.id.tv_current_language)
        cardUnits = findViewById(R.id.card_units)
        tvCurrentUnits = findViewById(R.id.tv_current_units)
        
        // Notifications
        switchPushNotifications = findViewById(R.id.switch_push_notifications)
        switchRecipeUpdates = findViewById(R.id.switch_recipe_updates)
        switchDailyFacts = findViewById(R.id.switch_daily_facts)

        // Privacy & Support
        cardPrivacy = findViewById(R.id.card_privacy)
        cardHelpSupport = findViewById(R.id.card_help_support)
        cardAbout = findViewById(R.id.card_about)
        
        // Account Actions
        cardLogout = findViewById(R.id.card_logout)
    }

    private fun loadCurrentSettings() {
        // Load theme
        val currentTheme = prefs.getString(KEY_THEME, THEME_DARK) ?: THEME_DARK
        tvCurrentTheme.text = when (currentTheme) {
            THEME_LIGHT -> "Light"
            THEME_DARK -> "Dark"
            THEME_SYSTEM -> "Follow System"
            else -> "Dark"
        }
        
        // Load language
        // Use LocaleHelper as the source of truth for language preference so the app-wide
        // locale and Settings view remain in sync even if other parts of the app write
        // to language preferences.
        val currentLanguage = LocaleHelper.getLanguage(this)
        // Keep a mirrored copy in this activity prefs for any other legacy code that
        // reads from `MixMateSettings`.
        prefs.edit().putString(KEY_LANGUAGE, currentLanguage).apply()
        tvCurrentLanguage.text = when (currentLanguage) {
            LANG_ENGLISH -> "English"
            LANG_AFRIKAANS -> "Afrikaans"
            LANG_ZULU -> "isiZulu"
            else -> "English"
        }
        
        // Load units
        val currentUnits = prefs.getString(KEY_UNITS, UNITS_METRIC) ?: UNITS_METRIC
        tvCurrentUnits.text = when (currentUnits) {
            UNITS_METRIC -> "Metric (ml, cl)"
            UNITS_IMPERIAL -> "Imperial (oz, fl oz)"
            else -> "Metric (ml, cl)"
        }
        
        // Load notification preferences
        switchPushNotifications.isChecked = prefs.getBoolean(KEY_PUSH_NOTIFICATIONS, true)
        switchRecipeUpdates.isChecked = prefs.getBoolean(KEY_RECIPE_UPDATES, false)
        switchDailyFacts.isChecked = prefs.getBoolean(KEY_DAILY_FACTS, false)
    }

    private fun setupClickListeners() {
        // Back button
        btnBack.setOnClickListener {
            finish()
        }
        
        // Account Section
        cardEditProfile.setOnClickListener {
            startActivity(Intent(this, EditProfileActivity::class.java))
        }
        
        cardChangePassword.setOnClickListener {
            showChangePasswordDialog()
        }
        
        // Preferences Section
        cardTheme.setOnClickListener {
            showThemeDialog()
        }
        
        cardLanguage.setOnClickListener {
            showLanguageDialog()
        }
        
        cardUnits.setOnClickListener {
            showUnitsDialog()
        }
        
        // Notifications
        switchPushNotifications.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
                    ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                    pendingEnablePush = true
                    requestGenericNotificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                } else {
                    NotificationHelper.createNotificationChannel(this)
                    prefs.edit().putBoolean(KEY_PUSH_NOTIFICATIONS, true).apply()
                    Toast.makeText(this, "Push notifications enabled", Toast.LENGTH_SHORT).show()
                }
            } else {
                prefs.edit().putBoolean(KEY_PUSH_NOTIFICATIONS, false).apply()
                Toast.makeText(this, "Push notifications disabled", Toast.LENGTH_SHORT).show()
            }
        }

        switchRecipeUpdates.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU &&
                    ContextCompat.checkSelfPermission(this, Manifest.permission.POST_NOTIFICATIONS) != PackageManager.PERMISSION_GRANTED) {
                    pendingEnableRecipe = true
                    requestGenericNotificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                } else {
                    NotificationHelper.createNotificationChannel(this)
                    prefs.edit().putBoolean(KEY_RECIPE_UPDATES, true).apply()
                    Toast.makeText(this, "Recipe updates enabled", Toast.LENGTH_SHORT).show()
                }
            } else {
                prefs.edit().putBoolean(KEY_RECIPE_UPDATES, false).apply()
                Toast.makeText(this, "Recipe updates disabled", Toast.LENGTH_SHORT).show()
            }
        }

        // Daily Cocktail Facts - Schedule notifications at 8 AM and 8 PM
        switchDailyFacts.setOnCheckedChangeListener { _, isChecked ->
            if (isChecked) {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                    when {
                        ContextCompat.checkSelfPermission(
                            this,
                            Manifest.permission.POST_NOTIFICATIONS
                        ) == PackageManager.PERMISSION_GRANTED -> {
                            NotificationHelper.createNotificationChannel(this)
                            enableDailyFacts()
                        }
                        shouldShowRequestPermissionRationale(Manifest.permission.POST_NOTIFICATIONS) -> {
                            // Show explanation and request permission
                            MaterialAlertDialogBuilder(this)
                                .setTitle(getString(R.string.notification_permission_rationale_title))
                                .setMessage(getString(R.string.notification_permission_rationale_message))
                                .setPositiveButton(getString(R.string.notification_permission_enable)) { _, _ ->
                                    requestNotificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                                }
                                .setNegativeButton(getString(R.string.notification_permission_not_now)) { _, _ ->
                                    switchDailyFacts.isChecked = false
                                }
                                .show()
                        }
                        else -> {
                            requestNotificationPermissionLauncher.launch(Manifest.permission.POST_NOTIFICATIONS)
                        }
                    }
                } else {
                    // Android 12 and below don't need runtime permission
                    enableDailyFacts()
                }
            } else {
                disableDailyFacts()
            }
        }

        // Privacy & Support
        cardPrivacy.setOnClickListener {
            showPrivacyPolicyDialog()
        }
        
        cardHelpSupport.setOnClickListener {
            showHelpSupportDialog()
        }
        
        cardAbout.setOnClickListener {
            showAboutDialog()
        }
        
        // Account Actions
        cardLogout.setOnClickListener {
            showLogoutConfirmation()
        }
    }

    // Theme Dialog
    private fun showThemeDialog() {
        val themes = arrayOf("Dark", "Light", "Follow System")
        val currentTheme = prefs.getString(KEY_THEME, THEME_DARK) ?: THEME_DARK
        val currentIndex = when (currentTheme) {
            THEME_DARK -> 0
            THEME_LIGHT -> 1
            THEME_SYSTEM -> 2
            else -> 0
        }
        
        MaterialAlertDialogBuilder(this)
            .setTitle("App Theme")
            .setSingleChoiceItems(themes, currentIndex) { dialog, which ->
                val selectedTheme = when (which) {
                    0 -> THEME_DARK
                    1 -> THEME_LIGHT
                    2 -> THEME_SYSTEM
                    else -> THEME_DARK
                }
                
                prefs.edit().putString(KEY_THEME, selectedTheme).apply()
                applyTheme(selectedTheme)
                tvCurrentTheme.text = themes[which]
                dialog.dismiss()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun applyTheme(theme: String) {
        when (theme) {
            THEME_DARK -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
            THEME_LIGHT -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_NO)
            THEME_SYSTEM -> AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM)
        }
    }

    // Language Dialog (POE Part 3 requirement)
    private fun showLanguageDialog() {
        val languages = arrayOf("English", "Afrikaans", "isiZulu")
        val currentLanguage = prefs.getString(KEY_LANGUAGE, LANG_ENGLISH) ?: LANG_ENGLISH
        val currentIndex = when (currentLanguage) {
            LANG_ENGLISH -> 0
            LANG_AFRIKAANS -> 1
            LANG_ZULU -> 2
            else -> 0
        }

        MaterialAlertDialogBuilder(this)
            .setTitle("Select Language")
            .setIcon(android.R.drawable.ic_dialog_info)
            .setSingleChoiceItems(languages, currentIndex) { dialog, which ->
                val selectedLanguage = when (which) {
                    0 -> LANG_ENGLISH
                    1 -> LANG_AFRIKAANS
                    2 -> LANG_ZULU
                    else -> LANG_ENGLISH
                }

                // Save to this activity prefs (legacy) and to the LocaleHelper prefs which
                // are used by attachBaseContext to apply the locale app-wide.
                prefs.edit().putString(KEY_LANGUAGE, selectedLanguage).apply()
                LocaleHelper.setLanguage(this, selectedLanguage)
                // Update UI immediately and recreate activity to apply new locale
                tvCurrentLanguage.text = languages[which]
                Toast.makeText(this, "Language updated", Toast.LENGTH_SHORT).show()
                dialog.dismiss()
                // Recreate activity so resources are reloaded with the new locale
                recreate()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    // Units Dialog
    private fun showUnitsDialog() {
        val units = arrayOf("Metric (ml, cl)", "Imperial (oz, fl oz)")
        val currentUnits = prefs.getString(KEY_UNITS, UNITS_METRIC) ?: UNITS_METRIC
        val currentIndex = if (currentUnits == UNITS_METRIC) 0 else 1
        
        MaterialAlertDialogBuilder(this)
            .setTitle("Measurement Units")
            .setSingleChoiceItems(units, currentIndex) { dialog, which ->
                val selectedUnits = if (which == 0) UNITS_METRIC else UNITS_IMPERIAL
                prefs.edit().putString(KEY_UNITS, selectedUnits).apply()
                tvCurrentUnits.text = units[which]
                Toast.makeText(this, "Units updated", Toast.LENGTH_SHORT).show()
                dialog.dismiss()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    // Change Password Dialog
    private fun showChangePasswordDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle("Change Password")
            .setMessage("This feature requires Firebase Authentication. You'll receive an email to reset your password.")
            .setPositiveButton("Send Email") { _, _ ->
                sendPasswordResetEmail()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun sendPasswordResetEmail() {
        val currentUser = FirebaseAuth.getInstance().currentUser
        val email = currentUser?.email ?: UserManager.getCurrentUserEmail(this)
        
        if (email.isNullOrEmpty()) {
            Toast.makeText(this, "No email found for this account", Toast.LENGTH_SHORT).show()
            return
        }
        
        FirebaseAuth.getInstance().sendPasswordResetEmail(email)
            .addOnSuccessListener {
                Toast.makeText(
                    this,
                    "Password reset email sent to $email",
                    Toast.LENGTH_LONG
                ).show()
            }
            .addOnFailureListener { e ->
                Log.e("SettingsActivity", "Failed to send password reset email", e)
                Toast.makeText(
                    this,
                    "Failed to send email: ${e.message}",
                    Toast.LENGTH_SHORT
                ).show()
            }
    }

    // Privacy Policy Dialog
    private fun showPrivacyPolicyDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle("Privacy Policy")
            .setMessage(
                "MixMate Privacy Policy\n\n" +
                "We respect your privacy. This app collects:\n\n" +
                "• Account information (email, username)\n" +
                "• Custom recipes you create\n" +
                "• Your favorite cocktails\n\n" +
                "Your data is stored securely in Firebase and is not shared with third parties.\n\n" +
                "For more information, visit our website."
            )
            .setPositiveButton("OK", null)
            .show()
    }

    // Help & Support Dialog
    private fun showHelpSupportDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle("Help & Support")
            .setMessage(
                "Need help with MixMate?\n\n" +
                "Email: support@mixmate.app\n" +
                "GitHub: github.com/mixmate/support\n\n" +
                "We typically respond within 24 hours."
            )
            .setPositiveButton("OK", null)
            .show()
    }

    // About Dialog
    private fun showAboutDialog() {
        MaterialAlertDialogBuilder(this)
            .setTitle("About MixMate")
            .setMessage(
                "MixMate v1.0.0\n\n" +
                "Your ultimate cocktail companion.\n\n" +
                "Developed by:\n" +
                "Teejay Kamwaro – ST10274142\n" +
                "Grant Coutts - ST10258297\n" +
                "Tristan Vries - ST10380906\n" +
                "Kelvin Gravett - ST10108660\n\n" +
                "© 2025 MixMate. All rights reserved."
            )
            .setPositiveButton("OK", null)
            .show()
    }

    // Logout Confirmation
    private fun showLogoutConfirmation() {
        MaterialAlertDialogBuilder(this)
            .setTitle("Logout")
            .setMessage("Are you sure you want to logout?")
            .setPositiveButton("Logout") { _, _ ->
                performLogout()
            }
            .setNegativeButton("Cancel", null)
            .show()
    }

    private fun enableDailyFacts() {
        prefs.edit().putBoolean(KEY_DAILY_FACTS, true).apply()
        NotificationScheduler.scheduleDailyNotifications(this)
        Toast.makeText(
            this,
            "Daily cocktail facts enabled! You'll receive notifications at 8 AM and 8 PM 🍹",
            Toast.LENGTH_LONG
        ).show()
    }

    private fun disableDailyFacts() {
        prefs.edit().putBoolean(KEY_DAILY_FACTS, false).apply()
        NotificationScheduler.cancelDailyNotifications(this)
        Toast.makeText(
            this,
            "Daily cocktail facts disabled",
            Toast.LENGTH_SHORT
        ).show()
    }

    private fun performLogout() {
        try {
            // Sign out from Firebase
            FirebaseAuth.getInstance().signOut()
            
            // Clear user data from UserManager
            UserManager.clearUserData(this)
            
            // Clear app-specific preferences if needed
            prefs.edit()
                .putBoolean(KEY_PUSH_NOTIFICATIONS, false)
                .putBoolean(KEY_RECIPE_UPDATES, false)
                .apply()
            
            Toast.makeText(this, "Logged out successfully", Toast.LENGTH_SHORT).show()
            
            // Navigate back to login
            val intent = Intent(this, MainActivity::class.java).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                putExtra("logout", true)
            }
            startActivity(intent)
            finish()
            
        } catch (e: Exception) {
            Log.e("SettingsActivity", "Error during logout", e)
            Toast.makeText(this, "Error logging out: ${e.message}", Toast.LENGTH_SHORT).show()
        }
    }
}
